### PR TITLE
Feature/base model

### PR DESCRIPTION
**Pull request by lkappeler**

### CODE REVIEW

File: src/LoopbackModelBase.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
+const MicroserviceError = require('./MicroserviceError');
+
+module.exports = class LoopbackModelBase {
+    constructor({ modelName, model }) {
```
Comment (rcky) [inline src/LoopbackModelBase.js:4]: Is there a reason to pass in an object? do we have any guidelines on this?

File: src/LoopbackModelBase.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        if (!modelName) {
+            throw LoopbackModelBase.createError(`No modelName was given to the
+                LoopbackModelBase constructor`);
+        }
+        if (!model) {
+            throw LoopbackModelBase.createError(`No Loopback Model Instance was given to the
+                LoopbackModelBase constructor`);
+        }
+
+        this.modelName = modelName;
```
Comment (rcky) [inline src/LoopbackModelBase.js]: We could just take the modelName of the loopback model if no other name was provided. In general I think it is a rare case you want to give our model a different name than the base model.

File: src/LoopbackModelBase.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        this[this.modelName] = model;
```
Comment (rcky) [inline src/LoopbackModelBase.js]: Don't see why you are assigning it to a variable property. It just makes to code less readable and might lead to collisions.

**Review by rcky — COMMENTED** [review]
I don't understand why the model is created this way. It'll create a model with a totally different interface than the basic loopback models (even though idk how to do it properly). Should we use mixins?

File: src/LoopbackModelBase.js (hunk 1 above)
Comment (linaGirl) [reply src/LoopbackModelBase.js:4]: i personally like it to use objects since it makes everything easy extendable

File: src/LoopbackModelBase.js [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+module.exports = class LoopbackModelBase {
+    constructor({ modelName, model }) {
+        if (!model) {
+            throw LoopbackModelBase.createError(`No Loopback Model Instance was given to the
+                LoopbackModelBase constructor`);
+        }
+
+        this.modelName = modelName || model.modelName;
+        this[this.modelName] = model;
```
Comment (linaGirl) [inline src/LoopbackModelBase.js]: why are you storing this under a dynamic name and not a static one? `this.loopbackModel` would be fine, not?

File: src/LoopbackModelBase.js [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+            if (!this[this.modelName][key]) {
+                this[this.modelName][key] = prototypeFunction.bind(this);
+            } else {
+                throw LoopbackModelBase.createError(`You are overwriting an loopback internal function ${key} on the model ${this.modelName}`);
+            }
+        });
+    }
+    
+    /**
+     * loops recursively trought the given objects pprototypes, reurns all
```
Comment (linaGirl) [inline src/LoopbackModelBase.js]: typos

File: src/LoopbackModelBase.js [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+     * prototoype functions except 'constructor' and the JavaScript Objects
+     * constructor prototype functions
+     *
+     * @param  {Object} object                  The object to extract prototype functions
+     * @param  {Map}    [functions=new Map()]   Map to store the functions
+     * @return {Map}                            A map with all prototpye functions
+     */
+    getPrototypeFunctionsRecursive(object, functions = new Map()) {
+        const prototype = Object.getPrototypeOf(object);
+        const isLastPrototype = !Object.getPrototypeOf(prototype);
```
Comment (linaGirl) [inline src/LoopbackModelBase.js:35]: Originally is was thinking of checking the name of the class, which is available under `instance.constructor.name` but i didn't remember where to find this property. The current solution works fine though.

File: src/LoopbackModelBase.js [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        this[this.modelName][hook](method, (...params) => {
+            const next = params.pop();
+
+            hookFunction.apply(this, [...params, registerContext])
+                .then((stop) => {
+                    if (!stop) {
+                        next();
+                    }
+                })
+                .catch((error) => {
```
Comment (linaGirl) [inline src/LoopbackModelBase.js]: could be `catch(next)`

File: src/LoopbackModelBase.js [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    /**
+     * Get the loopback environment
+     * @return {String} Environment variable
+     */
+    getEnv() {
+        if (!this.appEnv) {
+            if (!this[this.modelName].app) {
+                throw LoopbackModelBase.createError(`The loopback application environment could
```
Comment (linaGirl) [inline src/LoopbackModelBase.js]: What does this mean? What is the solution to this problem? Be more descriptive

**Review by linaGirl — CHANGES_REQUESTED** [review]

File: src/LoopbackModelBase.js (hunk 6 above)
Comment (lkappeler) [reply src/LoopbackModelBase.js:35]: ok

**Review by linaGirl — APPROVED** [review]

**Comment by rcky** [comment]
@eventEmitter @lkappeler This Merge broke the tests! Please make sure they run! I'll add the repos to the CI today. Also there are no tests for the Base Model!

**Pull request merged**